source/opt/loop_unroller.cpp

```diff
@@ -0,0 +1,704 @@
+
+  // Remove the final backedge to the header and make it point instead to the
+  // merge block.
+  state_.previous_continue_block_->tail()->SetInOperand(
+      0, {loop->GetMergeBlock()->id()});
+}
+
+// Uses the first loop to create a copy of the loop with new IDs.
+ir::Loop LoopUnrollerUtilsImpl::DuplicateLoop(ir::Loop* old_loop) {
+  const ir::Loop::BasicBlockOrderedListTy& basic_blocks =
+      old_loop->GetOrderedBlocks();
+
+  ir::Loop new_loop = *old_loop;
+
+  ir::Loop::BasicBlockOrderedListTy& new_block_order =
+      new_loop.GetOrderedBlocksRef();
+
+  new_block_order.clear();
+
+  // Copy every block in the old loop.
+  for (const ir::BasicBlock* itr : basic_blocks) {
+    CopyBasicBlock(old_loop, itr, true);
+    new_block_order.push_back(blocks_to_add_.back().get());
+  }
+
+  new_loop.SetLatchBlock(state_.new_continue_block);
+  new_loop.SetHeaderBlock(state_.new_header_block);
+
+  ir::BasicBlock* new_merge = old_loop->GetMergeBlock()->Clone(ir_context_);
+  AssignNewResultIds(new_merge);
+  state_.new_blocks[old_loop->GetMergeBlock()->id()] = new_merge;
+  new_loop.SetMergeBlock(new_merge);
+  for (auto& pair : state_.new_blocks) {
+    RemapOperands(pair.second);
+  }
+
+  return new_loop;
+}
+
+void LoopUnrollerUtilsImpl::AddBlocksToFunction(
+    const ir::BasicBlock* insert_point) {
+  for (auto basic_block_iterator = function_.begin();
+       basic_block_iterator != function_.end(); ++basic_block_iterator) {
+    if (basic_block_iterator->id() == insert_point->id()) {
+      basic_block_iterator.InsertBefore(&blocks_to_add_);
+      return;
+    }
+  }
+
+  assert(
+      false &&
+      "Could not add basic blocks to function as insert point was not found.");
+}
+
+// Assign all result_ids in |BB| instructions to new IDs and preserve the
+// mapping of new ids to old ones.
+void LoopUnrollerUtilsImpl::AssignNewResultIds(ir::BasicBlock* BB) {
+  // Label instructions aren't covered by normal traversal of the
+  // instructions.
+  uint32_t new_label_id = ir_context_->TakeNextId();
+
+  // Assign a new id to the label.
+  state_.new_inst[BB->GetLabelInst()->result_id()] = new_label_id;
+  BB->GetLabelInst()->SetResultId(new_label_id);
+
+  for (ir::Instruction& inst : *BB) {
+    uint32_t old_id = inst.result_id();
+
+    // Ignore stores ect.
+    if (old_id == 0) {
+      continue;
+    }
+
+    // Give the instruction a new id.
+    inst.SetResultId(ir_context_->TakeNextId());
+
+    // Save the mapping of old_id -> new_id.
+    state_.new_inst[old_id] = inst.result_id();
+
+    // Check if this instruction is the induction variable.
+    if (loop_induction_variable_->result_id() == old_id) {
+      // Save a pointer to the new copy of it.
+      state_.new_phi = &inst;
+    }
+  }
+}
+
+// For all instructions in |BB| check if the operands used are from a copied
+// instruction and if so swap out the operand for the copy of it.
+void LoopUnrollerUtilsImpl::RemapOperands(ir::BasicBlock* BB) {
```

Comment:
Good thing you did the AssignNewResultIds/RemapOperands, this will ease the def/use manager update